File: sim/golden_models/models/EmergencyShutdownModel.cpp
#include "golden_models/models/EmergencyShutdownModel.h"

#include "golden_models/core/TestVectorIO.h"

namespace fpd::sim {

void EmergencyShutdownModel::reset() {
    vgh_over_ = 0;
    vgh_under_ = 0;
    temp_over_ = 0;
    pll_unlocked_ = 0;
    hw_emergency_n_ = 1;
    shutdown_req_ = 0;
    force_gate_off_ = 0;
    shutdown_code_ = 0;
    cycle_count_ = 0;
}

void EmergencyShutdownModel::step() {
    shutdown_req_ = 0;
    force_gate_off_ = 0;
    shutdown_code_ = 0;
    if (hw_emergency_n_ == 0U) {
        shutdown_req_ = 1;
        force_gate_off_ = 1;
        shutdown_code_ = 0xEEU;
    } else if (vgh_over_ != 0U) {
        shutdown_req_ = 1;
        force_gate_off_ = 1;
        shutdown_code_ = 1;
    } else if (temp_over_ != 0U) {
        shutdown_req_ = 1;
        force_gate_off_ = 1;
        shutdown_code_ = 2;
    } else if (pll_unlocked_ != 0U) {
        shutdown_req_ = 1;
        force_gate_off_ = 1;
        shutdown_code_ = 3;
    } else if (vgh_under_ != 0U) {
        shutdown_req_ = 1;
        force_gate_off_ = 1;
        shutdown_code_ = 4;
    }
    ++cycle_count_;
}

void EmergencyShutdownModel::set_inputs(const SignalMap& inputs) {
    vgh_over_ = GetScalar(inputs, "vgh_over", vgh_over_);
    vgh_under_ = GetScalar(inputs, "vgh_under", vgh_under_);
    temp_over_ = GetScalar(inputs, "temp_over", temp_over_);
    pll_unlocked_ = GetScalar(inputs, "pll_unlocked", pll_unlocked_);
    hw_emergency_n_ = GetScalar(inputs, "hw_emergency_n", hw_emergency_n_);
}

SignalMap EmergencyShutdownModel::get_outputs() const {
    return {
        {"shutdown_req", shutdown_req_},
        {"force_gate_off", force_gate_off_},
        {"shutdown_code", shutdown_code_},
    };
}

std::vector<Mismatch> EmergencyShutdownModel::compare(const SignalMap& rtl_outputs) const {
    return CompareSignalMaps(cycle(), get_outputs(), rtl_outputs);
}

void EmergencyShutdownModel::generate_vectors(const std::string& output_dir) {
    TestVectorFile vectors;
    vectors.module_name = "emergency_shutdown";
    vectors.spec_name = "SPEC-FPD-008";
    vectors.clock_name = "sys_clk";
    vectors.signal_inputs = {"vgh_over", "vgh_under", "temp_over", "pll_unlocked", "hw_emergency_n"};
    vectors.signal_outputs = {"shutdown_req", "force_gate_off", "shutdown_code"};
    reset();
    set_inputs({{"vgh_over", 1U}, {"hw_emergency_n", 1U}});
    step();
    vectors.vectors.push_back({cycle(), {{"vgh_over", 1U}, {"hw_emergency_n", 1U}}, get_outputs()});
    WriteHexVectors(vectors, output_dir + "/emergency_vectors.hex");
    WriteBinaryVectors(vectors, output_dir + "/emergency_vectors.bin");
}

}  // namespace fpd::sim

File: sim/viewer/src/FpdSimViewer/Models/EmergencyShutdownModel.cs
using FpdSimViewer.Models.Core;

namespace FpdSimViewer.Models;

public sealed class EmergencyShutdownModel : GoldenModelBase
{
    private uint _vghOver;
    private uint _vghUnder;
    private uint _tempOver;
    private uint _pllUnlocked;
    private uint _hwEmergencyN = 1U;
    private uint _shutdownReq;
    private uint _forceGateOff;
    private uint _shutdownCode;

    public override void Reset()
    {
        _vghOver = 0U;
        _vghUnder = 0U;
        _tempOver = 0U;
        _pllUnlocked = 0U;
        _hwEmergencyN = 1U;
        _shutdownReq = 0U;
        _forceGateOff = 0U;
        _shutdownCode = 0U;
        CycleCount = 0;
    }

    public override void Step()
    {
        _shutdownReq = 0U;
        _forceGateOff = 0U;
        _shutdownCode = 0U;

        if (_hwEmergencyN == 0U)
        {
            _shutdownReq = 1U;
            _forceGateOff = 1U;
            _shutdownCode = 0xEEU;
        }
        else if (_vghOver != 0U)
        {
            _shutdownReq = 1U;
            _forceGateOff = 1U;
            _shutdownCode = 1U;
        }
        else if (_tempOver != 0U)
        {
            _shutdownReq = 1U;
            _forceGateOff = 1U;
            _shutdownCode = 2U;
        }
        else if (_pllUnlocked != 0U)
        {
            _shutdownReq = 1U;
            _forceGateOff = 1U;
            _shutdownCode = 3U;
        }
        else if (_vghUnder != 0U)
        {
            _shutdownReq = 1U;
            _forceGateOff = 1U;
            _shutdownCode = 4U;
        }

        CycleCount++;
    }

    public override void SetInputs(SignalMap inputs)
    {
        _vghOver = SignalHelpers.GetScalar(inputs, "vgh_over", _vghOver);
        _vghUnder = SignalHelpers.GetScalar(inputs, "vgh_under", _vghUnder);
        _tempOver = SignalHelpers.GetScalar(inputs, "temp_over", _tempOver);
        _pllUnlocked = SignalHelpers.GetScalar(inputs, "pll_unlocked", _pllUnlocked);
        _hwEmergencyN = SignalHelpers.GetScalar(inputs, "hw_emergency_n", _hwEmergencyN);
    }

    public override SignalMap GetOutputs()
    {
        return new SignalMap
        {
            ["shutdown_req"] = _shutdownReq,
            ["force_gate_off"] = _forceGateOff,
            ["shutdown_code"] = _shutdownCode,
        };
    }
}
